core/distributed/sparse_communicator.cpp

```diff
@@ -0,0 +1,275 @@
+ *
+ * \param exec the executor, this will always use the host executor
+ * \param comm communicator
+ * \param ids target ids of the one-sided operation
+ * \param sizes number of elements send to each id
+ *
+ * \return the inverse envelope consisting of the target-ids and the sizes
+ */
+std::tuple<array<comm_index_type>, std::vector<comm_index_type>>
+communicate_inverse_envelope(std::shared_ptr<const Executor> exec,
+                             mpi::communicator comm,
+                             const array<comm_index_type>& ids,
+                             const std::vector<comm_index_type>& sizes)
+{
+    auto host_exec = exec->get_master();
+    std::vector<comm_index_type> inverse_sizes_full(comm.size());
+    mpi::window<comm_index_type> window(host_exec, inverse_sizes_full.data(),
+                                        inverse_sizes_full.size(), comm,
+                                        sizeof(comm_index_type), MPI_INFO_ENV);
+    window.fence();
+    for (int i = 0; i < ids.get_size(); ++i) {
+        window.put(host_exec, sizes.data() + i, 1, ids.get_const_data()[i],
+                   comm.rank(), 1);
+    }
+    window.fence();
+
+    std::vector<comm_index_type> inverse_sizes;
+    std::vector<comm_index_type> inverse_ids;
+    for (int i = 0; i < inverse_sizes_full.size(); ++i) {
+        if (inverse_sizes_full[i] > 0) {
+            inverse_ids.push_back(i);
+            inverse_sizes.push_back(inverse_sizes_full[i]);
+        }
+    }
+
+    return std::make_tuple(
+        array<comm_index_type>{exec, inverse_ids.begin(), inverse_ids.end()},
+        std::move(inverse_sizes));
+}
+
+
+/**
+ * \brief
+ * \tparam LocalIndexType index type
+ * \param comm neighborhood communicator
+ * \param remote_local_idxs the remote indices in their local indexing
+ * \param recv_sizes the sizes that segregate remote_local_idxs
+ * \param send_sizes the number of local indices per rank that are part of
+ *                   remote_local_idxs on that ranks
+ * \return the local indices that are part of remote_local_idxs on other ranks,
+ *         ordered by the rank ordering of the communicator
+ */
+template <typename LocalIndexType>
+array<LocalIndexType> communicate_send_gather_idxs(
+    mpi::communicator comm, const array<LocalIndexType>& remote_local_idxs,
+    const array<comm_index_type>& recv_ids,
+    const std::vector<comm_index_type>& recv_sizes,
+    const array<comm_index_type>& send_ids,
+    const std::vector<comm_index_type>& send_sizes)
+{
+    // create temporary inverse sparse communicator
+    MPI_Comm sparse_comm;
+    MPI_Info info;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Info_create(&info));
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Dist_graph_create_adjacent(
+        comm.get(), send_ids.get_size(), send_ids.get_const_data(),
+        MPI_UNWEIGHTED, recv_ids.get_size(), recv_ids.get_const_data(),
+        MPI_UNWEIGHTED, info, false, &sparse_comm));
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Info_free(&info));
```

Comment:
Why not 
```suggestion
    GKO_ASSERT_NO_MPI_ERRORS(MPI_Dist_graph_create_adjacent(
        comm.get(), send_ids.get_size(), send_ids.get_const_data(),
        MPI_UNWEIGHTED, recv_ids.get_size(), recv_ids.get_const_data(),
        MPI_UNWEIGHTED, MPI_INFO_NULL, false, &sparse_comm));
```
since no key-value pairs are being set anyway ?

Reply:
One MPI implementation, I can't remember on which system, crashed without it, so I just kept it.